apps/web/src/services/thread-status.service.ts

```diff
@@ -0,0 +1,425 @@
+    return false;
+  }
+  const activePlanItems = getActivePlanItems(taskPlan);
+  return activePlanItems.every((planItem) => planItem.completed);
+}
+
+export class StatusResolver {
+  resolve(
+    manager: StatusResult,
+    planner?: StatusResult,
+    programmer?: StatusResult,
+  ): ThreadStatusData {
+    if (manager.status === "running" || manager.status === "error") {
+      return manager;
+    }
+
+    if (!planner) {
+      return manager;
+    }
+
+    if (planner.status === "running" || planner.status === "paused") {
+      return planner;
+    }
+
+    if (planner.status === "error") {
+      return planner;
+    }
+
+    if (!programmer) {
+      return planner;
+    }
+
+    return programmer;
+  }
+}
+
+const CACHE_TTL = 30 * 1000;
+
+function getCachedSessionData(
+  sessionCache: Map<string, any> | undefined,
+  sessionKey: string,
+): any | null {
+  if (!sessionCache) return null;
+
+  const cached = sessionCache.get(sessionKey);
+  if (!cached) return null;
+
+  const isExpired = Date.now() - cached.timestamp > CACHE_TTL;
+  if (isExpired) {
+    sessionCache.delete(sessionKey);
+    return null;
+  }
+
+  return cached;
+}
+
+function setCachedSessionData(
+  sessionCache: Map<string, any> | undefined,
+  sessionKey: string,
+  data: any,
+): void {
+  if (!sessionCache) return;
+
+  sessionCache.set(sessionKey, {
+    ...data,
+    timestamp: Date.now(),
+  });
+}
+
+export async function fetchThreadStatus(
+  threadId: string,
+  lastPollingState: ThreadStatusData | null = null,
+  managerThreadData?: Thread<ManagerGraphState> | null,
+  sessionCache?: Map<string, any>,
+): Promise<ThreadStatusData> {
+  try {
+    const apiUrl = process.env.NEXT_PUBLIC_API_URL ?? "";
+    if (!apiUrl) {
+      throw new Error("API URL not configured");
+    }
+
+    const client = createClient(apiUrl);
+    const resolver = new StatusResolver();
+
+    if (lastPollingState) {
+      try {
+        const optimizedResult = await checkLastKnownGraph(
+          client,
+          lastPollingState,
+          resolver,
+        );
+        if (optimizedResult) {
+          return optimizedResult;
+        }
+      } catch (error) {
+        console.warn(
+          "Optimization check failed, falling back to full status check:",
+          error,
+        );
+      }
+    }
+
+    return await performFullStatusCheck(
+      client,
+      threadId,
+      resolver,
+      managerThreadData,
+      sessionCache,
+    );
+  } catch (error) {
+    console.error(`Error fetching thread status for ${threadId}:`, error);
+
+    const graph = lastPollingState?.graph || "manager";
+    const runId = lastPollingState?.runId || "";
+    const errorThreadId = lastPollingState?.threadId || threadId;
+
+    return {
+      graph,
+      runId,
+      threadId: errorThreadId,
+      status: "error",
+    };
+  }
+}
+
+async function checkLastKnownGraph(
+  client: ReturnType<typeof createClient>,
```

Comment:
import from langgraph sdk
```suggestion
  client: Client,
```